### ISSUE TITLE
Lattice C modal dialogs lock up in xaaes 1.18

### ISSUE THREAD

**Issue by chrisridd** [freemint/freemint#185]
Hi,

I've upgraded from xaaes 1.17 (from AFROS) to xaaes 1.18.0 and have noticed a regression in form handling.

I use Lattice C 5.6, and when I open any of its modal dialogs in 1.18.0 the dialog draws but then none of the buttons work correctly. In 1.17 they worked OK.

I wrote a small app just doing a simple form_do() and noticed that xaaes 1.18 displayed it in a window. The Lattice modal dialogs are not appearing in windows so clearly Lattice is doing something extra.

Obviously Lattice C is (or was) a commercial product so I'm happy to provide additional log files if you can tell me what to enable etc.

**Comment by DavidGZ** [comment]
Hi,

Why didn't you upgrade to a more recent version (1.19)? If there is no particular reason, could you please try with a 1.19 snapshot?

https://freemint.github.io/#snapshots

**Comment by th-otto** [comment]
On Donnerstag, 4. Juni 2020 18:10:22 CEST Chris Ridd wrote:
> I wrote a small app just doing a simple form_do() and noticed that xaaes
> 1.18 displayed it in a window.

That should only happen when you don't put the form_do() call inside 
wind_update(BEG_MCTRL)/wind_update(END_MCTRL), which is required because 
form_do() needs control of the mouse.

**Comment by chrisridd** [comment]
> On Donnerstag, 4. Juni 2020 18:29:43 CEST DavidGZ wrote: Why didn't you upgrade to a more recent version (1.19)
> Ah, missed that. Yes, David is right, please try that with a recent version. We don't have the manpower to fix bugs in old versions, unless they are still present. Only the current version is actively maintained.

I totally agree, I just didn't realise the latest release here is not actually considered the latest release...

**Comment by DavidGZ** [comment]
> 
> 1.18 is the latest version on github.com! But I can certainly try a snapshot from github.io.

Yes, you're right sorry, a release should have been done for the 1.19, but we virtually don't have a lead maintainer so we live with the snapshots, but of course people who don't follow MiNT development or are new to MiNT don't know this.

**Comment by chrisridd** [comment]
OK, I just tried with 1-19-37f and Lattice C is still broken as described.

**Comment by th-otto** [comment]
Thats a bit strange. There are certainly a lot of programs using modal dialogs, and i've never heard of them being locked up. Could you eventually try that with your test program? Maybe it is just some other program (ACC or AUTO-folder) that causes that behaviour?

And also sorry for the flame about the version. The current source has been labeled 1.19 for several years now, but never officially released. Since automatic snapshots are available nobody felt the need to do an official release (but is has been discussed several times of course).

**Comment by chrisridd** [comment]
On 4 Jun 2020, at 20:01, Thorsten Otto <notifications@github.com> wrote:
> 
> ﻿
> Thats a bit strange. There are certainly a lot of programs using modal dialogs, and i've never heard of them being locked up. Could you eventually try that with your test program? Maybe it is just some other program (ACC or AUTO-folder) that causes that behaviour?
> 

I did have trouble migrating from 1.18 to 1.19 so it is certainly possible that something is interfering. I will try the stock drive_c from 1.19 and retest.

Migrating between releases seems pretty tricky...

> And also sorry for the flame about the version. The current source has been labeled 1.19 for several years now, but never officially released. Since automatic snapshots are available nobody felt the need to do an official release (but is has been discussed several times of course).
> 
No offence taken!

**Comment by skaftetryne** [comment]
On 04.06.2020 18:40, Thorsten Otto wrote:
> On Donnerstag, 4. Juni 2020 18:10:22 CEST Chris Ridd wrote:
>  > I wrote a small app just doing a simple form_do() and noticed that xaaes
>  > 1.18 displayed it in a window.
> 
> That should only happen when you don't put the form_do() call inside
> wind_update(BEG_MCTRL)/wind_update(END_MCTRL), which is required because
> form_do() needs control of the mouse.

Windowed form_do() is a great feature in XaAES, I used it in all the 
VanillaMiNT tools.

Jo Even

**Comment by chrisridd** [comment]
I've just tested with a clean drive_c from freemint-1-19-37f-040-aranym.zip and I'm seeing the same problem when Lattice runs a modal dialog. The dialog displays OK, I can click a few non-exit buttons in it, but as soon as I click an exit button the whole UI seems to wedge.

**Comment by th-otto** [comment]
I would consider that a bug in Lattice. It doesn seem to be very GEM-clean, there are for example some redraw problems:
![Screenshot_20200627_143059](https://user-images.githubusercontent.com/12425871/85922371-f3935600-b882-11ea-9692-f1c92fe8370b.png)
That looks to me as if it is unconditionally displaying it own info line, without looking at window rectangles.

Note that in XaAES, the submenus dropped by the menu are just normal windows. I think this has already been the case since at least 1.16

**Comment by mikrosk** [comment]
I gave this one an honest try, even hacked Hatari to print `intout` array (it normally prints only input parameters) to compare all the AES calls made. Basically LC5, when pressing OK in its About dialog, does a model loop like:

```C
form_center
  wind_update(END_UPDATE)
  form_dial(FMD_START, dialog_rect)
  objc_draw(dialog_rect)
  wind_update(END_MCTRL)
  loop:
      evnt_multi(MU_KEYBD | MU_BUTTON)            // modal: no MU_MESAG
      objc_find(x, y)         -> clicked object
      form_button(obj, clicks) -> (0, 0x26) on exit
      ... cleanup ...
```
The cleanup after form_button returns (0, 0x26) (i.e. exit detected) is where the two AES implementations diverge.

  The divergence
  
  TOS 4 (dialog closes):
  objc_find(0,8, 0x149,0x16a)   -> 0x26
  form_button(0x26, 0x1)         -> (0x0, 0x26)
  objc_edit(0,0, 0x3c40, ED_END) -> (0x1, 0x3c40)
  wind_update(BEG_MCTRL) 
  wind_update(BEG_UPDATE)
  form_dial(FMD_FINISH, rect)    -> (0x1)        ← closes dialog
  
  XaAES (dialog stays open):
  objc_find(0,8, 0x1FB,0x224)   -> 0x26
  form_button(0x26, 0x1)         -> (0x0, 0x26)
  objc_offset(0x26)              -> (0x1, 0x1E4, 0x211)
  objc_edit(0,0, 0xE000, ED_END) -> (0x1, 0xFFFF)
  objc_draw(0,8, 0x935B,0x8E3E,0x6F74,0x7406)  ← garbage clip rect!
  wind_update(END_MCTRL)                          ← END not BEG
  objc_edit(0x2, 0,0, ED_INIT)                    
  (back to evnt_multi)
